### ISSUE TITLE
loading more synthetic data that connects all aspects of pdc framework

### ISSUE THREAD

**Issue by jmergy** [PhilanthropyDataCommons/service#1382]
@slifty @kfogel as mentioned. I need to generate more synthetic data that will connect the various elements of the pdc framework capability for the Exchange development efforts to be better able to render layouts, variations in data, etc. Last time, I generated proposals and bulk uploaded but now I would want to have 5-10 opportunities, proposals, funders, changemakers, etc. I can use existing combos of the basefields we already have but I know a lot has changed over the last couple of months on structure so asking for some guidance. I can use the API to load but any sequence or preferred methodology would be helpful. Looking to do this later this week.

**Comment by bickelj** [comment]
@jmergy I now see you directed this toward @slifty and @kfogel but I will say that off the top of my head one thing to add is Sources for the various funders, changemakers, and data platform providers.

**Comment by jmergy** [comment]
@slifty @bickelj and @kfogel I am also trying to document a 5 or so step process to onboard new partners here and what they need to do or provide for us to get connected, add their organization reference, add basefields, add opportunities, and proposal data. I might just go through that to load more synthetic data. I believe the needs on all this have shifted from the original scramble of the pilot. I'm doing a draft and will ask for sanity checks. But, I also need to get connected data in ASAP for Exchange dev. Any guidance please.

**Comment by slifty** [comment]
The problem we have right now is that the bulk upload UX has not been updated to fully support sources -- it just uses the system source for all users.

This means you'd need to orchestrate bulk uploads via API calls.  It would look like this, done 5 times:

1) Create the separate bulk upload file using the template (and whatever base fields you want)
2) Create a new funder using the swagger interface
3) Create a new source using the swagger interface
4) Create a presigned post request using the swagger interface
5) Curl the bulk upload file to the presinged post request
6) Create a bulk upload task using the swagger interface + specifying the source created in step 2

**Comment by jmergy** [comment]
I will report back.

**Comment by jmergy** [comment]
Going to move through this in the next day or two @slifty might have questions/hit walls

**Comment by jmergy** [comment]
@slifty @bickelj as I start loading things. Just want to confirm a couple of areas. 

Sequence: Feels like I should perform the data loads in this sequence but correct me if I am wrong
1. Add baseFields (if needed) but I won't for this. I will use combos of what we have in there for now. 
2. Add funders - I will be adding 10 or so new funders 
3. Add opportunities - these will be combinations of basefields and connected to the funders from #2
4. Add sources - connected to the funders
5. Add changemakers - (if needed) but for this, I will just use the existing changemakers from the other synthetic data
6. Add proposals - linked to changemakers, opportunities, sources, and funders

Does this make sense in the new structures?

**Comment by bickelj** [comment]
@jmergy I think I would re-arrange/clarify it like this:

1. Add baseFields (if needed)
2. Add funders
3. Add sources - connected to the funders
4. Add opportunities
5. Add application forms - these will be combinations of basefields and connected to the opportunities from 3
6. Add changemakers - (if needed)
7. Add proposal versions - linked to changemakers and sources

If we update the front-end to allow source to be specified, the last two (or maybe more) steps can be achieved with bulk upload.

@slifty Does this sound right to you?

**Comment by jmergy** [comment]
Thanks @bickelj I think main goal is to get more diverse connected/linked data in ASAP so will proceed down that sequence if @slifty concurs with the thinking/approach here

**Comment by jmergy** [comment]
Got some funders in today. Will keep going.

**Comment by jmergy** [comment]
@bickelj @slifty I was able to get through everything up to adding changemakers. I don't think I need to. I can use the ones we have from the other synthetic data from 2024 for now for the proposals against the new funders, sources, opportunities, and application forms I just created. BUT, I am a little unclear how best to leverage the existing changemakers for the proposals. 

Questions:
1. Do I need to do anything to the existing changemakers in there now that I brought in back in 2024 because that was before the major changes to funders and changemakers, etc. Seems like I do need to do something. The sample changemaker record is different that the ones I brought in via bulk upload a while back. What should I do?
2. Should I try to use those changemakers from the bulk upload that might need some fixing or just do new ones and bring them in with the updated links? 
3. What is the connection of changemakers to organizations now? Is there one for funders too? I want to bring in funder address and org info but haven't yet. I did park funder records in there but just the name. 
4. Ideally, I want to bring in changemakers or connect existing changemakers in there now to new proposals I am generating against the various applicationForms and opportunities I just created. What am I missing here? I'm worried about bringing in proposals but not properly linking them to changemakers and organizations now. 

Thanks. This is all to get us some better, more connected data in PDC while we wait for real data.

**Comment by jmergy** [comment]
Quick update on status -

-     baseFields - in from past efforts
-     funders - added now (+10)
-     sources - added now and linked to funders (+10)
-     opportunities - added now and linked to funders "<fundername> 2025" convention (+10)
-     application forms - added with varying formats and groupings of baseFields and connected to opportunities (+10)

After I get some clarity on the above questions I can load or connecte changemakers and bring in varying proposals against applicationForms. Thanks

**Comment by slifty** [comment]
> Do I need to do anything to the existing changemakers in there now that I brought in back in 2024 because that was before the major changes to funders and changemakers, etc. Seems like I do need to do something. The sample changemaker record is different that the ones I brought in via bulk upload a while back. What should I do?

Changemakers created by bulk upload are the same as changemakers created by API, and any changes to changemakers have been retroactively applied -- old data was migrated to any new system (that said, not much changed about the entity other than the name of the entity I believe).

> Should I try to use those changemakers from the bulk upload that might need some fixing or just do new ones and bring them in with the updated links?

Changemakers created by bulk upload are the same type of entity as changemakers created by API.  The link is based on the name + EIN.

> What is the connection of changemakers to organizations now? Is there one for funders too? I want to bring in funder address and org info but haven't yet. I did park funder records in there but just the name.

There is no such thing as `organization` any longer.  That table was renamed to `changemaker` and we created two additional entity types `funder` and `data_provider`.  There are "organization fields" as opposed to "proposal fields" -- we may want to rename these to `changemaker fields` for consistency at some point.

> Ideally, I want to bring in changemakers or connect existing changemakers in there now to new proposals I am generating against the various applicationForms and opportunities I just created. What am I missing here? I'm worried about bringing in proposals but not properly linking them to changemakers and organizations now.

Relationships between changemakers and specific proposals are formed by POSTing to `/changemakerProposals` -- or by using the bulk upload.

**Comment by jmergy** [comment]
Thanks @slifty and I suppose my only issue now is getting in proposals that are connected to the applicationForm. I want to bring in data now for changemakers that is connected to specific applicationForms and the opportunity that is connected to, and the source that is connected to, and the funder that is connected to. 

I can't do that via the bulk upload right because I cann't specify the applicationForm correct? So, I am going to generate data related to a changemaker we already have in PDC so I can populate it via the API and it gets connected to the rest of the relevant connections. My understanding was bulk upload really just generates a new opportunity or applicationForm. That is not what I want to try and mimic the real data as much as possible. 

Is it correct to proceed on generating data for changemaker proposals and posting them with the applicationForm reference?

**Comment by slifty** [comment]
That's correct -- you'll need to make Proposals + proposal versions with proposal field values that reference the application form fields for the application you want the proposal to relate to.

**Comment by jmergy** [comment]
thanks I was able to get something and connected to the relevant other areas and will build off that! whoo hoo!
